### PR TITLE
feat(dev): add BrowserMeta file for @remix-run/dev

### PR DESCRIPTION
**Pull request by 19Qingfeng**
Closes: #6416 

- [ ] Docs
- [x] Tests

I opened up my may machine and ran this script:

1. Create a playgroud project, modify `metafile: true` in remix.config.js, and run `npm run build` to compile remix-run project.
After that, look at `debugDirectory`(`public/.debugger`) to get the `js-metafile.json` file.

2. Create a playgroud project, modify `metafile: true` and `debugDirectory: ".debugger"`  in remix.config.js, and run `npm run build` to compile remix-run project.
After that, look at `debugDirectory`(`.debugger`) to get the `js-metafile.json` file.

3. Create a playground project, execute `remix build --metafile`,then you will get `js-metafile.json`.

### CODE REVIEW

**Comment by pcattori** [comment]
@19Qingfeng thank you for the PR!

I think the direction we want to go in is to have a `--metafile` option that produces the metafile(s) into a dedicated directory.
Maybe default to `.debug/` and configurable with `debugDirectory` option in `remix.config.js` or something like that.

**Comment by 19Qingfeng** [comment]
> @19Qingfeng谢谢你的公关！
> 
> 我认为我们想要进入的方向是有一个`--metafile`选项可以将图元文件生成到一个专用目录中。 也许默认为`.debug/`并可使用`debugDirectory`选项 in`remix.config.js`或类似的东西进行配置。



> @19Qingfeng thank you for the PR!
> 
> I think the direction we want to go in is to have a `--metafile` option that produces the metafile(s) into a dedicated directory. Maybe default to `.debug/` and configurable with `debugDirectory` option in `remix.config.js` or something like that.

Okay, I will refer to your feedback to modify the parameter name and generation location. I will resubmit it later. Thank you.

**Comment by 19Qingfeng** [comment]
> @19Qingfeng thank you for the PR!
> 
> I think the direction we want to go in is to have a `--metafile` option that produces the metafile(s) into a dedicated directory. Maybe default to `.debug/` and configurable with `debugDirectory` option in `remix.config.js` or something like that.

@pcattori hello,pcattori.Now, this time with the metafile and debugDirectory parameters, I will update the test description later.

File: packages/remix-dev/config.ts [hunk 1]
```diff
@@ -268,6 +278,17 @@ export interface RemixConfig {
...
+  /**
+   * The absolute path to the metafile build directory.
+   */
+  debugDirectory: string
+
+  /**
+   * Whether to support generate browser js metafile for esbuild, default to false.
+   * When you set it to true, remix-js-metafile.json will be generated in the assetsBuildDirectory directory.
+   */
+  metafile?: boolean;
```
Comment (pcattori) [inline packages/remix-dev/config.ts:290]: `metafile` should be specifiable as a config option as you have here, but also as a flag `--metafile` for both `remix build` and `remix dev`.

File: packages/remix-dev/compiler/compiler.ts [hunk 2]
```diff
@@ -108,12 +113,12 @@ export let create = async (ctx: Context): Promise<Compiler> => {
     options.onManifest?.(manifest);
     writes.manifest = writeManifest(ctx.config, manifest);
 
-    // server compilation
```
Comment (pcattori) [inline packages/remix-dev/compiler/compiler.ts:111]: Could you restore this comment?

File: packages/remix-dev/compiler/utils/analyzer.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            fs.mkdirSync(debugDirectory)
+          }
+          return fs.writeFile(normalize(output), JSON.stringify(outputMeta));
+        } catch (e) {
+          invariant(e, `Failed to generate ${JS_META_FILE_NAME} in ${debugDirectory}.`);
+        }
+      }
+    }
+  }
+}
```
Comment (pcattori) [inline packages/remix-dev/compiler/utils/analyzer.ts:41]: Could you rewrite this as a function `writeMetafile( ctx: Context, name: string, metafile: Metafile)`?

File: packages/remix-dev/compiler/compiler.ts [hunk 4]
```diff
@@ -96,6 +100,7 @@ export let create = async (ctx: Context): Promise<Compiler> => {
     let js = await tasks.js;
     if (!js.ok) throw error ?? js.error;
     let { metafile, hmr } = js.value;
+    await createBrowserMetaFile(metafile)
```
Comment (pcattori) [inline packages/remix-dev/compiler/compiler.ts:103]: Conditional should be visible at this level:

```suggestion
    if (ctx.config.metafile) {
      await writeMetafile( ctx, "js", metafile)
    }
```

**Review by pcattori — CHANGES_REQUESTED** [review]

File: packages/remix-dev/compiler/compiler.ts (hunk 2 above)
Comment (19Qingfeng) [reply packages/remix-dev/compiler/compiler.ts:111]: Okay, this is really my oversight.

File: packages/remix-dev/compiler/compiler.ts (hunk 4 above)
Comment (19Qingfeng) [reply packages/remix-dev/compiler/compiler.ts:103]: All right, I'll take care of it later.

File: packages/remix-dev/compiler/compiler.ts (hunk 2 above)
Comment (19Qingfeng) [reply packages/remix-dev/compiler/compiler.ts]: In fact, a better solution is that I think metafile should be turned off by default under dev.

File: packages/remix-dev/config.ts (hunk 1 above)
Comment (19Qingfeng) [reply packages/remix-dev/config.ts:291]: Currently, metafile in my code will also be enabled according to the configuration when you use remix dev. But actually, I think maybe it's a good idea to shut down generate metafile all the time under remix dev.

**Comment by 19Qingfeng** [comment]
@pcattori hi, i updated the code and synchronized the description, please check the changes.

**Comment by 19Qingfeng** [comment]
@pcattori Can you provide some feedback? Perhaps I can make further modifications

**Comment by Doelfke** [comment]
What's the status of this?

**Comment by 19Qingfeng** [comment]
> What's the status of this?

I'm not sure what the remix team is thinking right now. If I have an idea, Maybe I can resolve the conflict and modify the code.

**Pull request closed without merging**